log-warper-core/src/Log/Core/Class.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE DefaultSignatures          #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | Monad type class for things that have logging ability.
+
+module Log.Core.Class
+       ( -- * Interface
+         MonadLogger (..)
+       , logR
+       , withR
+       , unliftLogAction
+
+         -- * Basic interface implementation
+       , ActionT (..)
+       , usingActionT
+       ) where
+
+import Control.Monad (void)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Class (MonadTrans (..))
+import Control.Monad.Trans.Except (ExceptT, mapExceptT, runExceptT)
+import Control.Monad.Trans.Identity (IdentityT (..), mapIdentityT)
+import Control.Monad.Trans.Maybe (MaybeT (..), mapMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask, mapReaderT)
+import Control.Monad.Trans.State.Strict (StateT (..), get, mapStateT)
+import Data.Kind (Type)
+
+import Log.Core.Action (LogAction (..), cmap)
+
+{- | Type class for monads that have logging ability. In order to log things
+monad @m@ should be able to construct 'LogAction' for some logging record @r@.
```

Comment:
What does "logging record" mean? This documentation is not clear. From the definition of `LogAction`, I now think it's the message type.

What about renaming this parameter to `message`? It's what `logging-effect` does: https://hackage.haskell.org/package/logging-effect-1.2.5/docs/Control-Monad-Log.html#t:MonadLog

In fact, what is the advantage of this `MonadLogger` over the one defined in `logging-effect`?

Reply:
> I now think it's the message type

Yes, my English is poor.

> What about renaming this parameter to message?

This sounds like a good improvement!

> In fact, what is the advantage of this MonadLogger over the one defined in logging-effect?

As I mentioned earlier, I never heard about `logging-effect` library. Need to think about this.